jujugui/static/gui/src/app/app.js

```diff
@@ -2079,6 +2000,55 @@ YUI.add('juju-gui', function(Y) {
         this.cookieHandler.check();
       }
       next();
+    },
+
+    /**
+      Takes a macaroon and stores the user info (if any) in the app.
+
+      @method _storeUser
+      @param {String} service The service the macaroon comes from.
+      @param {String} macaroon The base64 encoded macaroon.
+     */
+    _storeUser: function(service, macaroon) {
+      if (macaroon) {
+        var username = null;
+
+        // XXX jcsackett 2016-01-08: We should be using the whoami endpoint on
+        // jem to get the username of a logged in jem user rather than
+        // parsing the macaroon. Consider this all a bad hack for demo and we'll
+        // do something better in the very near future.
+        try {
+          macaroon = JSON.parse(atob(macaroon));
+          // We check all the caveats of each piece of the macaroon for the
+          // username data. If we find more than one username we abandon the
+          // macaroon as invalid.
+          macaroon.some(function(piece) {
+            var caveats = piece.caveats;
+            return caveats.some(function(caveat) {
+              var declaration = 'declared username ';
+              var index = caveat.cid.indexOf(declaration);
+              if (index !== -1) {
+                if (username === null) {
+                  username = caveat.cid.slice(index + declaration.length);
+                } else {
+                  console.log('Multiple usernames found in macaroon. ' +
+                              'Macaroon is invalid.');
+                  username = null;
+                  return true; // Breaks the some() function call.
+                }
+              }
+            });
+          });
+        } catch (e) {
+          console.log('Error working with macaroon json: ' + e);
+        }
+
+        if (username) {
+          this.set(service + 'User', {
```

Comment:
This is going to create attributes which are undocumented and impossible to grep for in the future. Basically we then have to make requests for random attributes which are not clear as to where they are from.

This should probably be an object of users keyed off of their service type:

```JavaScript
userDetails: {
  jem: { name: username},
  charmstore: { name: username},
  ...
}
```